File: packages/sdks/client-sdks/gatrix-js-client-sdk/src/FlagProxy.ts
/**
 * FlagProxy - Thin shell that delegates ALL logic to VariationProvider.
 *
 * Architecture per CLIENT_SDK_SPEC:
 * - Holds only flagName + forceRealtime + client reference.
 * - ALL property reads and variation methods delegate to the client.
 * - No deep copy of flag data — always reads live state from FeaturesClient cache.
 * - isRealtime property indicates the proxy's operational mode.
 * - Client is always present (never null).
 */
import { EvaluatedFlag, Variant, ValueType, VariationResult } from './types';
import { VariationProvider } from './VariationProvider';

export class FlagProxy {
  private _flagName: string;
  private _forceRealtime: boolean;
  private client: VariationProvider;

  constructor(client: VariationProvider, flagName: string, forceRealtime: boolean = false) {
    this._flagName = flagName ?? '';
    this._forceRealtime = forceRealtime;
    this.client = client;
  }

  // ==================== Properties ====================

  get name(): string {
    return this._flagName;
  }

  /** Whether this proxy was created in realtime mode. */
  get isRealtime(): boolean {
    return this._forceRealtime;
  }

  /** Whether the flag exists in the current cache. */
  get exists(): boolean {
    return this.client.hasFlagInternal(this._flagName, this._forceRealtime);
  }

  /**
   * Check if the flag is enabled.
   * Delegates to FeaturesClient for metrics tracking.
   */
  get enabled(): boolean {
    return this.client.isEnabledInternal(this._flagName, this._forceRealtime);
  }

  get variant(): Variant {
    return this.client.getVariantInternal(this._flagName, this._forceRealtime);
  }

  get valueType(): ValueType {
    return this.client.getValueTypeInternal(this._flagName, this._forceRealtime);
  }

  get version(): number {
    return this.client.getVersionInternal(this._flagName, this._forceRealtime);
  }

  get impressionData(): boolean {
    return this.client.getImpressionDataInternal(this._flagName, this._forceRealtime);
  }

  get raw(): EvaluatedFlag | undefined {
    return this.client.getRawFlagInternal(this._flagName, this._forceRealtime);
  }

  get reason(): string | undefined {
    return this.client.getReasonInternal(this._flagName, this._forceRealtime);
  }

  // ==================== Variation Methods ====================
  // All methods delegate to FeaturesClient's internal methods.
  // FlagProxy is a convenience shell - no own logic.

  variation(fallbackValue: string): string {
    return this.client.variationInternal(this._flagName, fallbackValue, this._forceRealtime);
  }

  boolVariation(fallbackValue: boolean): boolean {
    return this.client.boolVariationInternal(this._flagName, fallbackValue, this._forceRealtime);
  }

  stringVariation(fallbackValue: string): string {
    return this.client.stringVariationInternal(this._flagName, fallbackValue, this._forceRealtime);
  }

  numberVariation(fallbackValue: number): number {
    return this.client.numberVariationInternal(this._flagName, fallbackValue, this._forceRealtime);
  }

  jsonVariation<T>(fallbackValue: T): T {
    return this.client.jsonVariationInternal(this._flagName, fallbackValue, this._forceRealtime);
  }

  // ==================== Variation Details ====================

  boolVariationDetails(fallbackValue: boolean): VariationResult<boolean> {
    return this.client.boolVariationDetailsInternal(this._flagName, fallbackValue, this._forceRealtime);
  }

  stringVariationDetails(fallbackValue: string): VariationResult<string> {
    return this.client.stringVariationDetailsInternal(this._flagName, fallbackValue, this._forceRealtime);
  }

  numberVariationDetails(fallbackValue: number): VariationResult<number> {
    return this.client.numberVariationDetailsInternal(this._flagName, fallbackValue, this._forceRealtime);
  }

  jsonVariationDetails<T>(fallbackValue: T): VariationResult<T> {
    return this.client.jsonVariationDetailsInternal(this._flagName, fallbackValue, this._forceRealtime);
  }

  // ==================== Strict Variation Methods (OrThrow) ====================

  boolVariationOrThrow(): boolean {
    return this.client.boolVariationOrThrowInternal(this._flagName, this._forceRealtime);
  }

  stringVariationOrThrow(): string {
    return this.client.stringVariationOrThrowInternal(this._flagName, this._forceRealtime);
  }

  numberVariationOrThrow(): number {
    return this.client.numberVariationOrThrowInternal(this._flagName, this._forceRealtime);
  }

  jsonVariationOrThrow<T>(): T {
    return this.client.jsonVariationOrThrowInternal<T>(this._flagName, this._forceRealtime);
  }
}

File: packages/sdks/client-sdks/gatrix-flutter-sdk/lib/src/flag_proxy.dart
import 'models.dart';
import 'variation_provider.dart';

/// FlagProxy - Thin shell that delegates ALL logic to VariationProvider.
///
/// Architecture per CLIENT_SDK_SPEC:
/// - Holds only flagName + forceRealtime + client reference.
/// - ALL property reads and variation methods delegate to the client.
/// - No deep copy of flag data - always reads live state from FeaturesClient cache.
/// - isRealtime property indicates the proxy's operational mode.
/// - Client is always present (never null).
class FlagProxy {
  final VariationProvider _client;
  final String _flagName;
  final bool _forceRealtime;

  /// Primary constructor used by FeaturesClient internally.
  FlagProxy(this._client, this._flagName, {bool forceRealtime = false})
      : _forceRealtime = forceRealtime;

  /// Named-parameter constructor for test/mock use.
  /// [flag] is optional (null means flag doesn't exist).
  /// [client] is the VariationProvider to delegate to.
  /// [flagName] overrides the flag name (otherwise uses flag.name or '').
  FlagProxy.withFlag(
    EvaluatedFlag? flag, {
    required VariationProvider client,
    String? flagName,
    bool forceRealtime = false,
  })  : _client = client,
        _flagName = flagName ?? flag?.name ?? '',
        _forceRealtime = forceRealtime;

  // ==================== Properties ====================

  String get name => _flagName;

  /// Whether this proxy was created in realtime mode.
  bool get isRealtime => _forceRealtime;

  /// Whether the flag exists in the current cache.
  bool get exists =>
      _client.hasFlagInternal(_flagName, forceRealtime: _forceRealtime);

  /// Check if flag is enabled. Delegates to client for metrics tracking.
  bool get enabled =>
      _client.isEnabledInternal(_flagName, forceRealtime: _forceRealtime);

  Variant get variant =>
      _client.getVariantInternal(_flagName, forceRealtime: _forceRealtime);

  ValueType get valueType =>
      _client.getValueTypeInternal(_flagName, forceRealtime: _forceRealtime);

  int get version =>
      _client.getVersionInternal(_flagName, forceRealtime: _forceRealtime);

  bool get impressionData =>
      _client.getImpressionDataInternal(_flagName,
          forceRealtime: _forceRealtime);

  EvaluatedFlag? get raw =>
      _client.getRawFlagInternal(_flagName, forceRealtime: _forceRealtime);

  String? get reason =>
      _client.getReasonInternal(_flagName, forceRealtime: _forceRealtime);

  // ==================== Variation Methods ====================
  // All methods delegate to client's internal methods.
  // FlagProxy is a convenience shell - no own logic.

  String variation(String fallbackValue) => _client.variationInternal(
      _flagName, fallbackValue,
      forceRealtime: _forceRealtime);

  bool boolVariation(bool fallbackValue) => _client.boolVariationInternal(
      _flagName, fallbackValue,
      forceRealtime: _forceRealtime);

  String stringVariation(String fallbackValue) =>
      _client.stringVariationInternal(_flagName, fallbackValue,
          forceRealtime: _forceRealtime);

  int intVariation(int fallbackValue) => _client.intVariationInternal(
      _flagName, fallbackValue,
      forceRealtime: _forceRealtime);

  double doubleVariation(double fallbackValue) =>
      _client.doubleVariationInternal(_flagName, fallbackValue,
          forceRealtime: _forceRealtime);

  T jsonVariation<T>(T fallbackValue) => _client.jsonVariationInternal<T>(
      _flagName, fallbackValue,
      forceRealtime: _forceRealtime);

  // ==================== Variation Details ====================

  VariationResult<bool> boolVariationDetails(bool fallbackValue) =>
      _client.boolVariationDetailsInternal(_flagName, fallbackValue,
          forceRealtime: _forceRealtime);

  VariationResult<String> stringVariationDetails(String fallbackValue) =>
      _client.stringVariationDetailsInternal(_flagName, fallbackValue,
          forceRealtime: _forceRealtime);

  VariationResult<int> intVariationDetails(int fallbackValue) =>
      _client.intVariationDetailsInternal(_flagName, fallbackValue,
          forceRealtime: _forceRealtime);

  VariationResult<double> doubleVariationDetails(double fallbackValue) =>
      _client.doubleVariationDetailsInternal(_flagName, fallbackValue,
          forceRealtime: _forceRealtime);

  VariationResult<T> jsonVariationDetails<T>(T fallbackValue) =>
      _client.jsonVariationDetailsInternal<T>(_flagName, fallbackValue,
          forceRealtime: _forceRealtime);

  // ==================== Strict Variation Methods (OrThrow) ====================

  bool boolVariationOrThrow() => _client.boolVariationOrThrowInternal(
      _flagName,
      forceRealtime: _forceRealtime);

  String stringVariationOrThrow() => _client.stringVariationOrThrowInternal(
      _flagName,
      forceRealtime: _forceRealtime);

  int intVariationOrThrow() => _client.intVariationOrThrowInternal(_flagName,
      forceRealtime: _forceRealtime);

  double doubleVariationOrThrow() => _client.doubleVariationOrThrowInternal(
      _flagName,
      forceRealtime: _forceRealtime);

  T jsonVariationOrThrow<T>() => _client.jsonVariationOrThrowInternal<T>(
      _flagName,
      forceRealtime: _forceRealtime);
}
